File: app/src/main/java/com/door43/translationstudio/rendering/ClickableRenderingEngineFactory.java
package com.door43.translationstudio.rendering;

import android.content.Context;

import com.door43.translationstudio.core.TranslationFormat;
import com.door43.translationstudio.ui.spannables.Span;

/**
 * ClickableRenderingEngineFactory for creating ClickableRenderingEngine based on format
 */
public class ClickableRenderingEngineFactory {

    /**
     * create appropriate rendering engine for format and add click listeners
     * @param context
     * @param format
     * @param defaultFormat
     * @param verseClickListener
     * @param noteClickListener
     * @return
     */
    public static ClickableRenderingEngine create(
            Context context,
            TranslationFormat format,
            TranslationFormat defaultFormat,
            Span.OnClickListener verseClickListener,
            Span.OnClickListener noteClickListener
    ) {

        ClickableRenderingEngine renderer = null;

        if( (format != TranslationFormat.USFM) && (format != TranslationFormat.USX) ) {
            format = defaultFormat;
        }

        if(format == TranslationFormat.USFM) {
            renderer = new USFMRenderer(context, verseClickListener, noteClickListener);
        } if(format == TranslationFormat.USX)  {
            renderer = new USXRenderer(context, verseClickListener, noteClickListener);
        }

        return renderer;
    }

}

File: app/src/main/java/com/door43/translationstudio/rendering/ClickableRenderingEngineFactory.kt
package com.door43.translationstudio.rendering

import com.door43.translationstudio.core.TranslationFormat

/**
 * ClickableRenderingEngineFactory for creating ClickableRenderingEngine based on format
 */
object ClickableRenderingEngineFactory {

    /**
     * Create appropriate rendering engine for format using a verse display mode.
     * @param format
     * @param defaultFormat
     * @param verseDisplay how verse markers should be displayed
     * @return
     */
    fun create(
        format: TranslationFormat,
        defaultFormat: TranslationFormat,
        verseDisplay: VerseDisplay = VerseDisplay.NUMBER
    ): ClickableRenderingEngine {
        val resolvedFormat = if (format != TranslationFormat.USFM && format != TranslationFormat.USX) {
            defaultFormat
        } else {
            format
        }
        return when (resolvedFormat) {
            TranslationFormat.USFM -> USFMRenderer(verseDisplay = verseDisplay)
            TranslationFormat.USX -> USXRenderer(verseDisplay = verseDisplay)
            else -> throw IllegalArgumentException("Unsupported rendering format: $resolvedFormat")
        }
    }

}
